File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length(), max = 0;
        HashMap<Character, Integer> record = new HashMap<>();
        for (int i = 0, j = 0; i < len; i++) {
            char letter = s.charAt(i);
            if (record.containsKey(letter)) {
                j = Math.max(j, record.get(letter) + 1);
            }
            record.put(letter, i);
            max = Math.max(max, i - j + 1);
        }
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        unordered_map<int, int> mp;
        
        for (int l = 0, r = 0, n = s.size(); r < n; r++) {
            if (mp.find(s[r]) != mp.end()) {
                l = max(l, mp[s[r]] + 1);
            }
            
            mp[s[r]] = r;
            res = max(res, r - l + 1);
        }
        
        return res;
    }
};